<!-- move base -->
<launch>
  <master auto="start"/>

  <arg name="namespace"/>
  <arg name="configuration_basename" default="rrt_turtlebot3_lds_2d_gazebo.lua"/>

<!-- using cartographer -->
  <node pkg="cartographer_ros" type="cartographer_node" name="cartographer_node" 
        args="-configuration_directory $(find tdle)/param
              -configuration_basename $(arg configuration_basename)"
        output="screen">
  </node> 
  
  <node pkg="cartographer_ros" type="cartographer_occupancy_grid_node"
        name="cartographer_occupancy_grid_node" 
        args="-resolution 0.01" />

<!-- using Gmapping -->
    <!-- <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen" >
      <param name="map_frame" value="$(arg namespace)/map"/>
      <param name="odom_frame" value="$(arg namespace)/odom"/>
      <param name="base_frame" value="$(arg namespace)/base_footprint"/>
      <param name="map_update_interval" value="2.0"/>
      <param name="maxUrange" value="50.0"/>
      <param name="maxRange" value="50.0"/>
      <param name="sigma" value="0.05"/>
      <param name="kernelSize" value="1"/>
      <param name="lstep" value="0.05"/>
      <param name="astep" value="0.05"/>
      <param name="iterations" value="5"/>
      <param name="lsigma" value="0.075"/>
      <param name="ogain" value="3.0"/>
      <param name="lskip" value="0"/>
      <param name="srr" value="0.01"/>
      <param name="srt" value="0.02"/>
      <param name="str" value="0.01"/>
      <param name="stt" value="0.02"/>
      <param name="linearUpdate" value="0.01"/>
      <param name="angularUpdate" value="0.01"/>
      <param name="temporalUpdate" value="0.1"/>
      <param name="resampleThreshold" value="0.5"/>
      <param name="particles" value="30"/>
      <param name="xmin" value="-5.0"/>
      <param name="ymin" value="-5.0"/>
      <param name="xmax" value="5.0"/>
      <param name="ymax" value="5.0"/>
      <param name="delta" value="0.1"/>
      <param name="llsamplerange" value="0.01"/>
      <param name="llsamplestep" value="0.01"/>
      <param name="lasamplerange" value="0.005"/>
      <param name="lasamplestep" value="0.005"/>
      <param name="minimumScore" value="0.005"/>
      <remap from='/odom' to ='/odometry/filtered' /> 
    </node> -->

  <node pkg="move_base" type="move_base" respawn="false" name="move_base_node" output="screen">
    <param name="footprint_padding" value="0.01" />
    <param name="recovery_behaviors" value="false" />
    <param name="controller_frequency" value="5.0" />
    <param name="controller_patience" value="3.0" />
    <param name="oscillation_timeout" value="30.0" />
    <param name="oscillation_distance" value="0.5" />
    <param name="planner_patience" value="1" />
    <param name="controller_patience" value="1" /> 
    <!-- <remap from="cmd_vel" to="mobile_base/commands/velocity"/> -->
    <param name="recovery_behavior_enabled" value="false" />
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>

    <rosparam file="$(find tdle)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find tdle)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find tdle)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find tdle)/param/global_costmap_params.yaml" command="load" />
    <!-- <rosparam file="$(find tdle)/param/base_local_planner_params.yaml" command="load" />  -->
    <rosparam file="$(find tdle)/param/dwa_local_planner_params_waffle.yaml" command="load" />
    
    <param name="global_costmap/global_frame" value="$(arg namespace)/map"/>
    <param name="global_costmap/robot_base_frame" value="$(arg namespace)/base_link"/>
  
    <param name="global_costmap/laser_scan_sensor/sensor_frame" value="$(arg namespace)/base_scan"/>
    <param name="global_costmap/laser_scan_sensor/topic" value="$(arg namespace)/scan"/>    
    <param name="local_costmap/global_frame" value="$(arg namespace)/odom"/>
    <param name="local_costmap/robot_base_frame" value="$(arg namespace)/base_link"/>
  
    <param name="local_costmap/laser_scan_sensor/sensor_frame" value="$(arg namespace)/base_scan"/>
    <param name="local_costmap/laser_scan_sensor/topic" value="/$(arg namespace)/scan"/>
    <param name="local_costmap/obstacle_layer/laser_scan_sensor/topic" value="/$(arg namespace)/scan"/>

  </node>

</launch>